### DISCUSSION TITLE
Ollama purges model from VRAM despite OLLAMA_KEEP_ALIVE=-1

### DISCUSSION THREAD

**Discussion by SD-fan** [open-webui/open-webui#3291] [Q&A] [answered]
I'm successfully runing open-webui with Ollama backend. 

Unfortunately when using open-webUI with **each** prompt the model loaded in Ollama gets purged from VRAM and is re-loaded to evaluate the prompt. This takes quite some time, rendering open-webUI useless for chatting.  

There is plenty of VRAM to keep the model running. I've set keep_alive in the settings to -1 as well as set the env variable OLLAMA_KEEP_ALIVE=-1. Unfortunately model still gets purged.

I'm also using Ollama as backend for an embedding model (simple RAG) and suspect there might be an issue/bug keeping both models (embedding and LLM) in VRAM? 

Anyone else experiencing the issue? Any work-around?

**Comment by atgehrhardt** [comment]
Wait have you set the OLLAMA_NUM_PARALLEL environmental variable?

By default Ollama will only load one model at a time. If you want to override that and allow two in memory at once, you need to add

OLLAMA_NUM_PARALLEL=2

If you haven't done this and you're using ollama for embeddings and chat it unloading and reloading each model every time.

**Reply by SD-fan** [reply]
@atgehrhardt Thanks for your swift reply. Unfortunately setting OLLAMA_NUM_PARALLEL=2 did not solve the issue. The model still gets purged with every prompt.

**Reply by atgehrhardt** [reply]
Hmmm very interesting. To my knowledge openwebui isn't exerting any influence over how models load in Ollama so this may actually be a better question for them, but I'm wondering if it's related to the environment.

Is it running on host or under docker?

**Reply by SD-fan** [reply]
@atgehrhardt All is running under docker. I've used the image with integrated Ollama using this 
`docker run -d -p 3000:8080 --env=OLLAMA_KEEP_ALIVE=-1 --env=OLLAMA_NUM_PARALLEL=2 --gpus=all -v ollama:/root/.ollama -v open-webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:ollama` 

I've also just pulled a new image. Unfortunately the issue persists and the model get's unloaded with each prompt still. 

I'll try to run Ollama in a separate instance as a work-around or do you have other ideas?

**Reply by justinh-rahb** [reply, marked as answer]
If you're using a custom `num_ctx` value or a different task model to what you're using for the chat then you'll get model swapping.

`OLLAMA_NUM_PARALLEL` controls how many concurrent generations occur, but to load more than one model you need to use `OLLAMA_MAX_LOADED_MODELS`

**Reply by SD-fan** [reply]
@justinh-rahb Awesome! Setting OLLAMA_MAX_LOADED_MODELS=2 did the trick. Now OpenWebUI is usable and I'm very grateful. Thanks so much for the great work 👍

**Comment by alanqoudif** [comment]
Hi, I have a question and need someone to explain it to me. I'm working on a project using Docker. I cloned it from GitHub, made some modifications, and added some features. After finishing the modifications, I uploaded it to my GitHub account. But the problem is, when I create a container to run the project, it doesn't seem to recognize the changes I made and goes back to running the original project from GitHub. And when I use the repository URL of my project, I get a message saying I need to upload it to Docker Hub. Honestly, I didn't fully understand the process. So, can I rely on you for that? I'll give you the repository URL and I want you to make the necessary adjustments to the repository, deliver the changes to me, and help me run the commands on our servers. Which package is suitable for this? Please note that the project is ready; I just need to add the modifications, upload it to Docker, and run it on our server. Additionally, if you prove to be helpful, I will work with you on two similar projects.
I updated the code and tried the result but the updates did not appear after the update

I just want to add the modifications and upload them to Docker, and then I will complete the upload process on my VPS server 
it,s her in my repo https://github.com/alanqoudif/Nuqta-web-ui

https://hub.docker.com/r/dyrnq/open-webui

**Reply by justinh-rahb** [reply]
No support for forks

**Reply by alanqoudif** [reply]
hi i don't understand
